share/nixos-shell/nixos-shell.nix

```diff
@@ -1,56 +1,156 @@
-  '';
-
-  systemd.services."serial-getty@ttyS0".enable = true;
-  networking.firewall.enable = false;
-
-  imports = lib.optional (nixos_config != "") nixos_config;
+
+  config = let
+    user = builtins.getEnv "USER";
+    shell = builtins.baseNameOf (builtins.getEnv "SHELL");
+  in lib.mkMerge [
+    # Enable the module of the user's shell for some sensible defaults.
+    (lib.mkIf (options.programs ? ${shell}.enable) {
+      programs.${shell}.enable = mkVMDefault true;
+    })
+
+    (lib.mkIf (pkgs ? ${shell}) {
+      users.extraUsers.root.shell = mkVMDefault pkgs.${shell};
+    })
+
+    (let
+      home = builtins.getEnv "HOME";
+    in lib.mkIf (home != "" && cfg.mounts.mountHome) {
+      users.extraUsers.root.home = lib.mkVMOverride home;
+    })
+
+    # Allow passwordless ssh login with the user's key if it exists.
+    (let
+      pubkey = "${builtins.getEnv "HOME"}/.ssh/id_rsa.pub";
+    in lib.mkIf (builtins.pathExists pubkey) {
+      users.users.root.openssh.authorizedKeys.keyFiles = [ pubkey ];
+    })
+
+    {
+      # Allow the user to login as root without password.
+      users.extraUsers.root.initialHashedPassword = "";
+
+      services.mingetty.helpLine = ''
+        Log in as "root" with an empty password.
+      '';
+
+      virtualisation = {
+        graphics = mkVMDefault false;
+        memorySize = mkVMDefault "500M";
+
+        qemu.options = let
+          nixProfile = "/nix/var/nix/profiles/per-user/${user}/profile/";
+        in
+          lib.optional cfg.mounts.mountHome "-virtfs local,path=/home,security_model=none,mount_tag=home" ++
+          lib.optional (cfg.mounts.mountNixProfile && builtins.pathExists nixProfile) "-virtfs local,path=${nixProfile},security_model=none,mount_tag=nixprofile" ++
+          lib.mapAttrsToList (target: mount: "-virtfs local,path=${builtins.toString mount.target},security_model=none,mount_tag=${mount.tag}") cfg.mounts.extraMounts;
+      };
+
+      # build-vm overrides our filesystem settings in nixos-config
+      boot.initrd.postMountCommands =
+        (lib.optionalString cfg.mounts.mountHome ''
+          mkdir -p $targetRoot/home/
+          mount -t 9p home $targetRoot/home/ -o trans=virtio,version=9p2000.L,cache=${cfg.mounts.cache}
+        '') +
+        (lib.optionalString (user != "" && cfg.mounts.mountNixProfile) ''
+          mkdir -p $targetRoot/nix/var/nix/profiles/per-user/${user}/profile/
+          mount -t 9p nixprofile $targetRoot/nix/var/nix/profiles/per-user/${user}/profile/ -o trans=virtio,version=9p2000.L,cache=${cfg.mounts.cache}
+        '') +
+        builtins.concatStringsSep " " (lib.mapAttrsToList (target: mount: ''
+          mkdir -p $targetRoot/${target}
+          mount -t 9p ${mount.tag} $targetRoot/${target} -o trans=virtio,version=9p2000.L,cache=${mount.cache}
+        '') cfg.mounts.extraMounts);
+
+      environment = {
+        systemPackages = with pkgs; [
+          xterm # for resize command
```

Comment:
Not in this pr, but this would shell
script would be probably more lightweight: 

https://github.com/nix-community/nixos-generators/blob/master/formats/vm-nogui.nix#L3